src/fixate/core/switching.py

```diff
@@ -0,0 +1,743 @@
+
+            mux_c = ("a3_c0", "a3_c1", "a3_c2", None)
+            mux_b = ("a1_b0", "a1_b1", "a1_b2", None)
+
+            map_tree = (
+                "a0",
+                mux_b,
+                "a2",
+                shift_nested(mux_c, [2]),  # 2 in indicative on how many pins to skip. This case is (x2, x3) from mux_b
+            )
+
+        Note:
+            If shift_nested is needed, I think I'd re-implement something like TreeMap that
+            can be used to define the mux.
+
+        class TreeMap:
+            signals: Sequence[pin | TreeMap]
+            pins: PinSet
+
+        class Mux(VirtualMux):
+            mux_c = TreeMap(("a3_c0", "a3_c1", "a3_c2", None), ("x4", "x5"))
+            mux_b = TreeMap(("a1_b0", "a1_b1", "a1_b2", None), ("x2", "x3"))
+            map_tree = TreeMap(("a0", mux_b, "a2", mux_c), ("x1", "x0"))
+        """
+        signal_map: SignalMap = dict()
+
+        bits_at_this_level = (len(tree) - 1).bit_length()
+        pins_at_this_level = pins[:bits_at_this_level]
+
+        for signal_or_tree, pins_for_signal in zip(
+            tree, generate_bit_sets(pins_at_this_level)
+        ):
+            if signal_or_tree is None:
+                continue
+            if isinstance(signal_or_tree, Signal):
+                signal_map[signal_or_tree] = frozenset(pins_for_signal) | fixed_pins
+            else:
+                signal_map.update(
+                    self._map_tree(
+                        tree=signal_or_tree,
+                        pins=pins[bits_at_this_level:],
+                        fixed_pins=frozenset(pins_for_signal) | fixed_pins,
+                    )
+                )
+
+        return signal_map
+
+    def __repr__(self) -> str:
+        return f"{self.__class__.__name__}('{self._state}')"
+
+    @staticmethod
+    def _default_update_pins(
+        pin_updates: PinUpdate, trigger_update: bool = True
+    ) -> None:
+        """
+        Output callback to effect a state change in the mux.
+
+        This is a default implementation which simply prints the planned state change to.
+        stdout. When instantiated as part of a jig driver, this will end up connected
+        to an AddressHandler to do the actual pin changes in hardware.
+
+        In general, this method shouldn't be overridden in a subclass. An alternative
+        can be provided to __init__.
+        """
+        print(pin_updates, trigger_update)
+
+
+class VirtualSwitch(VirtualMux):
+    """
+    A VirtualMux that controls a single pin.
+
+    A virtual switch is a multiplexer with a single pin. The multiplex
+    function can accept either boolean values or the strings 'TRUE' and
+    'FALSE'. The virtual address used to switch can be defined as a list
+    with the single element (as if it were a multiplexer) or by using the
+    shorthand which is to define the pin_name attribute as a string.
+    """
+
+    pin_name: Pin = ""
+    map_tree = ("FALSE", "TRUE")
```

Comment:
@jcollins1983 @daniel-montanari I'm think of changing this. Most scripts that use a `VirtualSwitch` pass in a bool literal. It seems silly have signal names that are exactly the same. I'm inclined to either do away with the `"TRUE"` and `"FALSE"`. Or turn them into `"On"` and `"Off"`.

Any strong thoughts?

(if it's not obvious, I'm at the point now where there is going to be enough work to migrate existing scripts that we might as well clean up any cruft we can in the process)

Reply:
I've made this change, but happy to roll it back if there is consensus.

Reply:
Not super fussed which way we do things. I just don't like having two ways of doing something.
ELV scripts are the main ones still using "TRUE" and "FALSE"

Reply:
I'm happy to keep the change. As far as "switches" go, most people probably think of them in terms of on and off not true and false, unless they're a logic nut ;p